### PR TITLE
Unit test conversion and TGD modernization (WIP)

### PR DESCRIPTION
**Pull request by dodexahedron**
This is a work-in-progress for conversion of unit tests for TGD to NUnit 4.0, as well as expansion, improvement, and modernization of various tests.
This also includes some preliminary incidental modernization work on TGD itself, primarily aimed at eliminating some reflection, and some redundant operations.

See https://github.com/dodexahedron/TerminalGuiDesigner for current activity

### CODE REVIEW

**Comment by dodexahedron** [comment]
I want to call particular attention to https://github.com/dodexahedron/TerminalGuiDesigner/issues/28 and its associated branch.

ViewFactory didn't have any explicit separate tests defined, so I'm making some for it.

Since it's a pretty important core component, I've started with some tests that are not functional unit tests, but rather change control tests, to enforce very intentional changes to its core functionality. For example, if a new type becomes supported and is removed from the known unsupported types that are excluded, at least one of these tests would fail, unless the same change is made in the test itself. This keeps unintended side-effects of various types of changes from creeping in and creating new bugs or regressions.

I'll of course add normal tests, once I'm done with those, but I've made some initial commits to the branch associated with that issue that result in 24 test cases (per target framework version), for the change control enforcement on ViewFactory.

As part of it, I've also been making some fundamental changes to the way ViewFactory works (primarily the Create method, which is the meat of it - I've created a generic version now). However, the changes result in the same output, which I've been careful to ensure (while also adding minor flexibility that might be handy in the future, such as additional parameters for dimensions). These changes are non-breaking, and I've left the old method there, for the time being, with an [Obsolete] attribute (and it internally delegates to the new method, for types I've implemented in it so far). I have also updated usages of `Create(Type)` in TGD to use the new `Create<T>()` method, one at a time, while verifying tests continue to pass in their existing form. Then I also make the tests call the generic method, once that is verified. I also (so far) moved a constant from another class into ViewFactory, to make a cleaner dependency graph.

If you want to peruse the changes, I do make a lot of micro-commits, so it is generally pretty easy to follow individual changes. But I'm more than willing to discuss it all directly (and kinda prefer that anyway).

Apologies for any style differences. That isn't me being passive-aggressive or anything haha. It's mostly just whitespace differences (and also explicit types on declarations, usually). Both are not a big deal to change on my end, though, if I have a spec you'd like me to follow. The stylecop rules were really not compatible with newer language constructs and were extremely noisy, so I disabled them in my local config. StyleCop seems to be sorta-kinda-dead, anyway?\
ReSharper is doing the formatting, on my end, using settings that _mostly_ follow MS style guides (with deviations both to adhere a _little_ closer to this project's original code, as well as some personal preferences that are mostly just for readability over all else). I'm of course more than willing to work out ReSharper formatting settings that will bring it closer to what was previously in the project. Either way, I can also share the resulting team-shared DotSettings files, if you like. Bottom line on that is it isn't my intent to impose style changes, if they are disliked, and I'll go back and work on changing it, if you want. :)

 That branch hasn't yet been merged to the branch in this PR, but it will be once I'm done writing the ViewFactory tests.

File: src/ViewFactory.cs [hunk 1]
```diff
@@ -13,46 +11,122 @@ namespace TerminalGuiDesigner;
 /// sensible dimensions and content for dragging/configuring in
 /// the designer.
 /// </summary>
-public class ViewFactory
+public static class ViewFactory
```
Comment (tznind) [inline src/ViewFactory.cs:14]: I normally don't like creating static classes as it can create limitations in future e.g. it prevents inheritence.

File: src/ViewFactory.cs [hunk 2]
```diff
@@ -13,46 +11,122 @@ namespace TerminalGuiDesigner;
...
+            case TableView tv:
+                var dt = new DataTable( );
+                dt.Columns.Add( "Column 0" );
+                dt.Columns.Add( "Column 1" );
+                dt.Columns.Add( "Column 2" );
+                dt.Columns.Add( "Column 3" );
+                tv.Width = width ?? 50;
+                tv.Height = height ?? 5;
+                tv.Table = new DataTableSource( dt );
+                break;
```
Comment (tznind) [inline src/ViewFactory.cs:91]: Love the new generic method it is very cool and easy to use in tests.  But I don't like that there is now 2 implementations of the code to instantiate each class.

We need to get rid of the duplication.

Easiest way to do so would be to centralize all the logic in the new generic and call it from the old version e.g.

```csharp
public static View Create(Type t)
{
        MethodInfo method = typeof(ViewFactory)
            .GetMethods(BindingFlags.Static | BindingFlags.Public)
            .Single(m=>
            m.IsGenericMethod && m.Name == nameof(ViewFactory.Create));

        MethodInfo generic = method.MakeGenericMethod(t);
        return (View)generic.Invoke(null, new object[] { null,null });
}
```

File: tests/ColorSchemeTests.cs [hunk 3]
```diff
@@ -121,26 +119,30 @@ public void TestColorSchemeProperty_ToString(bool testMultiSelectingSeveralTimes
...
         }
 
         selection.SetSelection(p.Design);
-        ClassicAssert.AreNotEqual(pink, p.Design.View.ColorScheme, "Expected view to be selected so be green not pink");
-        ClassicAssert.AreEqual("ColorScheme:pink", p.ToString(), "Expected us to know it was pink under the hood even while selected");
+        Assert.Multiple( ( ) =>
+        {
+            Assert.That( p.Design.View.ColorScheme, Is.Not.EqualTo( pink ), "Expected view to be selected to be green, not pink");
+            Assert.That( p.ToString(), Is.EqualTo( "ColorScheme:pink" ), "Expected us to know it was pink under the hood even while selected");
+        } );
```
Comment (tznind) [inline tests/ColorSchemeTests.cs:130]: Love these changes!

File: tests/CopyPasteTests.cs [hunk 4]
```diff
@@ -65,46 +71,68 @@ public void CopyPasteTableView()
         var copy = new CopyOperation(tvDesign);
         OperationManager.Instance.Do(copy);
 
-        ClassicAssert.AreEqual(0, OperationManager.Instance.UndoStackSize,
-            "Since you cannot Undo a Copy we expected undo stack to be empty");
+        // TODO: Remove this comment once addressed
+        // I think it would be a good idea to make the undoStack internal and then
+        // expose internals to the test project namespace, so we can actually inspect
+        // the collection itself.
+        // Using this proxy property only gives the test limited validity.
```
Comment (tznind) [inline tests/CopyPasteTests.cs:78]: I agree that this would allow more comprehensive testing.  I will create an issue.

File: src/ToCode/CodeDomArgs.cs [hunk 5]
```diff
@@ -69,6 +69,7 @@ public CodeDomArgs(CodeTypeDeclaration rootClass, CodeMemberMethod initMethod)
     /// so should have a capital letter at start. <see langword="false"/> for private members
     /// which should start with a lower case letter.</param>
     /// <returns><paramref name="name"/>.</returns>
+    /// TODO: Consider prefixing @ for reserved keywords, so casing rules can be preserved
```
Comment (tznind) [inline src/ToCode/CodeDomArgs.cs:72]: I considered this when I wrote this method but decided against it.

`ICodeGenerator` doesn't distinguish between 'reserved keywords' and simply invalid e.g. starting with a number.

Rather than try to list all the reserved type names myself I went with the simplest solution which was to treat all invalid names the same way which is to prepend with `_`.  

For example:

```
// Valid
public int @new {get;set;}

// Not Valid
public int @32 {get;set;}
```

File: src/UI/Windows/BigListBox.cs [hunk 6]
```diff
@@ -50,39 +49,30 @@ public class BigListBox<T>
     public BigListBox(
         string prompt,
         string okText,
-        bool addSearch,
+        in bool addSearch,
```
Comment (tznind) [inline src/UI/Windows/BigListBox.cs:52]: Is there a reason why we need the `in` keyword for this method?

File: src/UI/Windows/Modals.cs [hunk 7]
```diff
@@ -97,6 +98,44 @@ internal static bool GetArray(string windowTitle, string entryLabel, Type arrayE
         return false;
     }
 
+    internal static bool TryGetArray<T>(string windowTitle, string entryLabel, Array? initialValue, out Array? result)
```
Comment (tznind) [inline src/UI/Windows/Modals.cs:101]: I don't like that there are now 2 copies of very similar code.  It is nice to have generic method but needs refactoring so that there are as few duplicated lines as possible.

Simplest way to do that would be for the generic to call the specific e.g.

```csharp
internal static bool TryGetArray<T>(string windowTitle, string entryLabel, Array? initialValue, out Array? result)
{
    return GetArray(windowTitle, entryLabel, typeof(T), initialValue, out result);
}
```

Also lets rename both to be called `TryGetArray`

**Comment by tznind** [comment]
Thanks for your hard work on this.  Love the changes! I have left a few comments, let me know if any of them require elaboration.

File: src/ViewFactory.cs (hunk 2 above)
Comment (dodexahedron) [reply src/ViewFactory.cs:91]: I've got improvements to this already in the pipeline that haven't been pushed yet. Stay tuned!

Though I am discovering some edge cases related to run-time compilation and certain things that may either prevent full use of generics or at least require additional work, which I'm investigating at the moment.

File: tests/ColorSchemeTests.cs (hunk 3 above)
Comment (dodexahedron) [reply tests/ColorSchemeTests.cs:130]: I like the fluet-like way NUnit went, too. Makes them read like a sentence, for nice and easy review and maintenance. Kinda hits you over the head with exactly what it's doing. 😅

File: tests/CopyPasteTests.cs (hunk 4 above)
Comment (dodexahedron) [reply tests/CopyPasteTests.cs:78]: I've been doing it anyway, as I've been writing some tests, so you might not have to bother.

Just flipping them to internal and then adding the internalsvisibleto in the project file was all I had to do.

File: src/ToCode/CodeDomArgs.cs (hunk 5 above)
Comment (dodexahedron) [reply src/ToCode/CodeDomArgs.cs:72]: As it turns out, I got a list of all illegal keywords and the associated unit test now has them all. 🙂

But yes, there are some interesting limitations with the roslyn stuff, and also some potentially useful parts we might benefit from, as well.

File: src/UI/Windows/BigListBox.cs (hunk 6 above)
Comment (dodexahedron) [reply src/UI/Windows/BigListBox.cs:52]: Need? No. Small efficiency gain, and callers don't have to specify it, so it's a non-breaking change.

File: src/UI/Windows/Modals.cs (hunk 7 above)
Comment (dodexahedron) [reply src/UI/Windows/Modals.cs:101]: A agree completely.

However, it will require a LOT of work, as that will require a bunch of non-generic code to be genericized.

So, since the API isn't really intended for external consumption, I thought it would be acceptable to do it in a more gradual way. During that transition, there would be both, but the end goal is no duplication, for sure. Does that make sense?

**Comment by dodexahedron** [comment]
Love the feedback.

Gave a few quick responses, but I'm on my phone on my way to a Christmas party, so apologies if they're a bit incomplete. 

There's a lot of work beyond this that hasn't been pushed yet, because I've discovered some things that spidered out quite a bit that I'm wanting to be sure I address as well and as close to the original intent as possible.

File: src/ViewFactory.cs (hunk 1 above)
Comment (dodexahedron) [reply src/ViewFactory.cs:14]: Valid concern. Everything here was static and it's a factory, so it made sense at the time. But it can easily be non-static, since the methods are static.

It just wasn't necessary to instantiate it at any place it was used.

Easy enough to drop static on the class itself. Wouldn't require any other changes.

File: src/ViewFactory.cs (hunk 2 above)
Comment (dodexahedron) [reply src/ViewFactory.cs:91]: Oh also. The duplication is all gone in my current code. The old method now has a single switch that just proxies to the new one, withiut having its own implementation of any of them. That allowed me to move on with more pressing work withiut having to convert the entire application to generics right this second. 😅

File: src/UI/Windows/Modals.cs (hunk 7 above)
Comment (dodexahedron) [reply src/UI/Windows/Modals.cs:101]: Just to add to my previous response, the current implementation limits just how far I can take generics, since most of the rest isn't generic and passes Types around. That approach is actually somewhat _easier_, ironically, around some of the reflection, but of course has its costs, or at least can cause at least a little bit of code duplication, though most or all of that can be avoided by using proxies that do the Type-based checks and then call the generic for the actual implementation. But, it makes testing more difficult, and the reflection wouldn't be necessary with a pure generic implementation, anyway.

I've got a bunch of other thoughts, too - all of which tie in - which could have profound impacts on the ease of maintaining and improving TGD, in the future, but they're all definitely their own work items.

File: src/UI/Windows/BigListBox.cs (hunk 6 above)
Comment (dodexahedron) [reply src/UI/Windows/BigListBox.cs:52]: Expanding on that...

I generally try to put `in` on all value type parameters. There's no downside and it really should have been default behavior. At the native level, it avoids one memory copy in all situations, and, in cases where the parameter is never used by the callee for assignment to anything else (like just being used for an if statement or something), the value is never copied at all.

It also allows the compiler to make additional optimizations, in some cases, since it's a mathematical restriction on the program, as a whole.

One neat thing you gain from it, for free, is that you're prohibited from attempting to mutate the parameter, when it's a value type, because in means the reference cannot be reassigned, and changing a value type means changing that. So it prevents a subtle source of potential bugs since, otherwise, you can operate on the parameter, but, being a value type, you've copied it and are now no longer touching the original.

The fact that callers don't have to specify `in` makes it even nicer, since it still gets called that way anyway. It's optional at the call site.

I honestly wish c# treated value type parameters that way by default.

Plus, if you ever switch to a more complex value type, in its place, you already have it there and reap significant benefits from not passing a large struct around by value.

In another project I created, I went back and changed most value type parameters to `in` (and some immutable ref types, too, like strings) and the results were tangible to a degree I did not expect, both in memory working set of the application and in CPU usage. The unit tests even sped up noticeably.

File: src/ToCode/CodeDomArgs.cs (hunk 5 above)
Comment (dodexahedron) [reply src/ToCode/CodeDomArgs.cs:72]: So, looking at this on a larger screen...

I actually prefer prepending with an underscore, anyway, since that follows the .net design guide on private member naming and is what I'm used to in other projects. I actually don't remember why I had that thought when I wrote the TODO haha.

Comment (dodexahedron) [reply src/ToCode/CodeDomArgs.cs:72]: Oh yeah! I had another thought about this.

For the same reason I suggested making various things in TGD internal (testing), we should probably have it emit the members as internal, as well, instead of private. I've had to turn a lot of them into internal, manually, when using TGD, so that I can test, but they of course get turned back on next edit.

Maybe we could expose accessibility of members as a choice for the user?

File: src/ViewFactory.cs (hunk 2 above)
Comment (dodexahedron) [reply src/ViewFactory.cs:91]: Side-note about type checks.

Switch statements have some nice advantages, here, especially in generics.

Visual studio is smart enough to see if you have checks in the wrong order, which would result in an earlier check catching something because it's less-derived than a more specific one below. Plus, they're nice and compact, which is a cool side-benefit.

But yeah, once I push the commit that includes the cleanup of the old method, I think you'll like it. It's simple, duplicates nothing, doesn't have to reflect, and keeps existing code working without requiring us to change everything all at once.

**Comment by dodexahedron** [comment]
I've also got quite a bit of running commentary on the various issues I've filed for things, over on my fork, for the work items related to all of this.

Some of it is just me putting my thoughts in writing, but some of it is also exploring real-world problems, opportunities for improvements, and my thinking regarding changes I make.

File: src/ViewFactory.cs (hunk 2 above)
Comment (dodexahedron) [reply src/ViewFactory.cs:91]: Oooo another thought:

Once generics are more ubiquitous, throughout the project, the Create method (and thus the ViewFactory, in general) will actually be possible to remove all together, with its functionality just contained in the constructor of the future generic Design class.

**Comment by dodexahedron** [comment]
So...

Just to keep some of the more general, yet still single-topic-ish stuff a bit more contained, I turned on discussions on my fork and opened these two:

Generics:\
https://github.com/dodexahedron/TerminalGuiDesigner/discussions/32

Code Generation:\
https://github.com/dodexahedron/TerminalGuiDesigner/discussions/33

Nothing there yet, but just wanted to get ahead of things before we have a hundred scattered comments we can't keep track of. 😅

**Comment by dodexahedron** [comment]
So...

Now that I'm restricting the changes to just test conversion and necessary code to support that, I'd say you can play with merging this to whatever working branch you have, as you want, because I'm only merging completed and working units of work to this branch.

I just finished with DimExtensionsTests a little while ago, and that's now been merged to this branch, as well.

With the improvements, expansions, and unrolling of tests, so far, just in the few that have been done, we're now up to 1130 test cases (565 per framework version).

In DimExtensionsTests, I did some additional stuff I didn't _have_ to do, mostly to serve as additional examples of ways to use newer NUnit capabilities/constructs (though they actually reduce total code, vs not doing it that way, anyway, so yay).

**Comment by dodexahedron** [comment]
A comment on two test methods, which are very similar to the originals, in DimExtensionsTests:

The last couple that test the ToCode method only tested two DimTypes, in the original.

Was that intentional, or should I add a case to test the remaining type? I would think that I should, even if the test is just to prove a known failure case or something, or else the code isn't actually fully tested.

**Comment by dodexahedron** [comment]
Actually... Thinking about that last question...

Only one is actually necessary, for a definitively tight set of tests - the ToCode without offset (which also doesn't exist for Absolute).

So long as it is proven to return the expected code for that case, testing it in the offset case is redundant, since Terminal.Gui is responsible for that and we just get back a new absolute dim with the value added to it.

**Comment by dodexahedron** [comment]
Hey question about the Operation class and its relation to the OperationManager that came up as I'm working on the EditorTests fixture:

The `Do` method on an operation is not equivalent to the `OperationManager.Do()` method.

Is this intentional?

Seems to me that one should probably delegate to the other, so that they have the same behavior.

The impact of the current model is that calling Do() directly on an operation does not invoke the change tracking behavior that OperationManager.Do() has.

But, that change tracking is also dependent on the undoStack being modified, which means change tracking in theory does not work unless the last action that was executed had SupportsUndo == true, and either its explicitly modifies the undoStack itself or was invoked via the OperationManager, that operation will not be considered a change, for the unsaved change tracking.

I would suggest that unsaved change tracking be modified to use something not dependent on an operation supporting undo.

A change I added during work on a previous text fixture added execution counting to operations. Currently, it only increments, on Do(), but the ultimate intention was to make it also decrement, if undo was ever run on it. It was added to enable a stronger test guarantee, as it only increments if Operation.Do() is about to return true.

I have a really simple event-based solution for it that I can write up to show you (I'll stick it in a separate branch).
It's not much work to do, and would solve these problems, so I'm happy to show at least a prototype, whether you want to use it or not.

**Comment by dodexahedron** [comment]
Bah

Actually, no. That's me letting myself get side-tracked again.

I'll table that for later, and will file an issue on my fork so I don't forget.

**Comment by dodexahedron** [comment]
Filed https://github.com/dodexahedron/TerminalGuiDesigner/issues/34 to keep track of that for later. Back to working on EditorTests for now.

**Comment by dodexahedron** [comment]
EditorTests is converted, and ObjectExtensionsTests is converted, with new tests added.

I've got somewhere to be, so I'll continue work on adding more tests to ObjectExtensionsTests later, but I merged current work for visibility.

**Comment by tznind** [comment]
> Hey question about the Operation class and its relation to the OperationManager that came up as I'm working on the EditorTests fixture:
> 
> The `Do` method on an operation is not equivalent to the `OperationManager.Do()` method.
> 
> Is this intentional?

Yes, being able to `Do` an operation without having to put it on the undo list is useful.  

e.g. if you want to have a composite operation that does multiple other sub operations and don't want them ramming themselves onto the undo stack.  But you still want to be able to have those sub operations atomically undoable.


> The impact of the current model is that calling Do() directly on an operation does not invoke the change tracking behavior that OperationManager.Do() has.

This is intenional

> 
> But, that change tracking is also dependent on the undoStack being modified, which means change tracking in theory does not work unless the last action that was executed had SupportsUndo == true, and either its explicitly modifies the undoStack itself or was invoked via the OperationManager, that operation will not be considered a change, for the unsaved change tracking.

I would like to see a unit test demonstrating this.  Typically I think things that do not support undo don't get pushed into undo history.  Some things like Copy cannot sensibly be undone (or are otherwise unexpected e.g. in word you do not expect Ctrl+Z to edit your clipboard).

> 
> I would suggest that unsaved change tracking be modified to use something not dependent on an operation supporting undo.

I'm open to all ideas.

> A change I added during work on a previous text fixture added execution counting to operations. Currently, it only increments, on Do(), but the ultimate intention was to make it also decrement, if undo was ever run on it. It was added to enable a stronger test guarantee, as it only increments if Operation.Do() is about to return true.
> 
> I have a really simple event-based solution for it that I can write up to show you (I'll stick it in a separate branch). It's not much work to do, and would solve these problems, so I'm happy to show at least a prototype, whether you want to use it or not.

Would love to see this.  It may be you have a better design (given the above constraints).

**Comment by dodexahedron** [comment]
> > Hey question about the Operation class and its relation to the OperationManager that came up as I'm working on the EditorTests fixture:
> > The `Do` method on an operation is not equivalent to the `OperationManager.Do()` method.
> > Is this intentional?
> 
> Yes, being able to `Do` an operation without having to put it on the undo list is useful.
>
> e.g. if you want to have a composite operation that does multiple other sub operations and don't want them ramming themselves onto the undo stack. But you still want to be able to have those sub operations atomically undoable.
> 
> > The impact of the current model is that calling Do() directly on an operation does not invoke the change tracking behavior that OperationManager.Do() has.
> 
> This is intenional


Coolio. I will keep that in mind with whatever I come up with, then, to maintain those capabilities.

 
> > But, that change tracking is also dependent on the undoStack being modified, which means change tracking in theory does not work unless the last action that was executed had SupportsUndo == true, and either its explicitly modifies the undoStack itself or was invoked via the OperationManager, that operation will not be considered a change, for the unsaved change tracking.
> 
> I would like to see a unit test demonstrating this. Typically I think things that do not support undo don't get pushed into undo history. Some things like Copy cannot sensibly be undone (or are otherwise unexpected e.g. in word you do not expect Ctrl+Z to edit your clipboard).

Yeah. The IsImpossible and SupportsUndo, in the current implementation, handle that just fine. My goal would be to keep the same behavior, while just making testing a bit easier. And potentially moving the intelligence into the application itself, so you can always call Do and just expect it to `Do` it properly. Which I think is even already possible with what you've already got in there, too, without any major refactoring. But, for now, it is definitely something for me to watch for in tests.

> > I would suggest that unsaved change tracking be modified to use something not dependent on an operation supporting undo.
> 
> I'm open to all ideas.

👌

As above, I'd like to keep true to your original intent. I don't want to step on toes - just provide help where I can, with opportunistic enhancements, mostly, especially right now. 🙂

What follows is somewhat stream of consciousness writing what pops into my head, so you've been warned (also, that means it's likely to have lots of holes) 😅: \ 
So my main "master plan," if you want to call it that, for such an undertaking, if it is ever done, involves a potential reduction in the size and complexity of the operation class hierarchy, by them mostly being data necessary to undo, redo, or repeat an action, and by separating some common pieces in the interface and base class into, for example, an ISupportsUndo and ISupportsRedo interface, so only the ones that need it get it, and events make all the magic work together. Events would handle a lot of the meat of what they currently do, really, and whichever way the subscription works would be what determines where the code actually lives. They're actually a good candidate to be records or even record structs, depending on a million things between now and then.

It's not really a change in user-facing functionality - it's simply a change in the execution method/programming paradigm, from synchronous imperative function dispatch to synchronous and asynchronous (not necessarily meaning `async` mind you, though that's cool too) event-based dispatch. Same code, different places, less unneeded code for operations that don't support things in the first place (and that also can tie into unsaved change and undo/redo tracking, as well, and makes implementing things like INotifyPropertyChanged trivial).

Obviously, to have the functionality that exists, the code, or at least what it does, in some form, has to be _somewhere_, and I imagine a fair amount would end up in the OperationManager, with only the specific functionality related to each one living inside it, as an event handler. I'd have to actually whiteboard it out or prototype some code to figure that all out, since sometimes control is better inverted and sometimes not. One big bonus would be a likely improvement to thread safety, almost for free, since the singleton itself wouldn't be responsible for managing the events - they themselves would, via the events, and the singleton basically just becomes a service. That could even be tossed into the "modern" .net "service" infrastructure (all DI-like), with almost no modification, from that point.

> > A change I added during work on a previous text fixture added execution counting to operations. Currently, it only increments, on Do(), but the ultimate intention was to make it also decrement, if undo was ever run on it. It was added to enable a stronger test guarantee, as it only increments if Operation.Do() is about to return true.
> > I have a really simple event-based solution for it that I can write up to show you (I'll stick it in a separate branch). It's not much work to do, and would solve these problems, so I'm happy to show at least a prototype, whether you want to use it or not.
> 
> Would love to see this. It may be you have a better design (given the above constraints).

Well, it was a preliminary step toward future ideas. Partially just to keep focused on getting the tests done, I didn't go farther, and made only non-impacting changes. Baby steps (plus, that means way less likely to run into merge conflicts as you're working on other stuff). You can see it in the current code in this branch, in the base Operation class (and IOperation, though nothing uses it via the interface at the moment). It's just a `private int` field, a get-only `internal ref int` property pointing to that field, and thread-safe incrementing of the field, in the Do method, right before it's about to return true. Its use is limited to just execution counting, such as in tests, at the moment (which was the immediate goal anyway), so tests can prove that the same instance of an Operation hasn't run more than once when it shouldn't have, even though other references, such as to views themselves are maintained (relevant especially for copy/paste).

Later on, decrementing in the Undo method would be the simplest next step. Or, for more robust tracking possibilities, it could just be a separate counter, and then 3 total properties: one exposing Do/Redo count, one exposing Undo count, and one for convenience that just returns the difference of the two.

The rest of the design is quite a bit to try to convey in this form, but nothing is set in stone as far as I'm concerned, and I still have plenty to familiarize myself with, in the application, as I'm doing the test work, which will probably impact every one of these thoughts, before it's time to actually act on them.

Partially related, keeping tighter track of things would even open up an easier possibility of supporting "repeatable" actions, while still keeping reliable undo history, for example.

Anyway, enough of my rambling haha. That's all down the road, since I'm focusing on tests right now, which will take some time, at the pace I can dedicate time to it. I'm sure there will be more thoughts and questions, along the way. 😅

We're up to 573 tests per framework, as of the last code I merged to this branch. Onward!

**Comment by dodexahedron** [comment]
While I'm at it, is there anything in particular, with the unit test project, that you've been wanting?

File: src/ViewFactory.cs (hunk 2 above)
Comment (tznind) [reply src/ViewFactory.cs:99]: > Oh also. The duplication is all gone in my current code. The old method now has a single switch that just proxies to the new one, withiut having its own implementation of any of them.

I would like to merge this branch ASAP because it has lots of great changes.

Currently only outstanding issue for me is this 2 implementations in `ViewFactory`.  Is it possible to cherry pick your fix for that into this branch? Or I can redirect the method calls myself (but don't want to create merge conflicts if possible).

**Comment by dodexahedron** [comment]
KeyMapTests has been converted.

Along with this came a small move toward standardization for configuration, using Microsoft.Extensions.Configuration.

It's still YAML and still uses YamlDotNet under the covers, but now we have the capability to add literally three lines to allow users to use any format they like for configuration of keymaps.

KeyMap and ColorSchemeBlueprint became `record` classes, as part of this.

They're configuration classes, so rigorous testing of value equality is prudent. Options were to either manually implement (and then have to maintain, in perpetuity) equality methods and operators, and their associated tests, or simply change them to records, which get value equality for free.

I did go a little farther and went ahead and made a primary constructor for them, as well, just for flexibility and for the brevity it allows in the uses of them in the existing code, so I suppose they also gained that.

Tests for KeyMap now do the following:

 - Check that the default parameterless constructor does, in fact, give the expected values.
 - A change control test, to ensure the test yaml file isn't changed without the test and code also being updated to match.
 - Checks that loading a yaml file produces the expected KeyMap.
 - Checks that reference and value equality behave as they should.
 - Checks that non-destructive mutation (another cool thing gained for free with records) behaves as expected, with respect to value and reference equality of the KeyMap instances, as well as the mutated properties

To ensure someone can't accidentally break the KeyMap tests, there are also updated metadata files for git and VS that enforce LF line endings, for test purposes, and in the IDE, in the context of the KeyMapTests.cs file and the stock Keys.yaml file. The configuration test also ensures that hasn't been changed accidentally. This does not affect the actual behavior or requirements of the TerminalGuiDesigner application, itself. It just removes the need to do mutations on the string data in the tests, as was done before, just to ensure line endings don't break them.

We are now up to 629 tests per framework - all passing.

**Comment by dodexahedron** [comment]
oops. Didn't mean to close that. 🤦‍♂️

**Comment by dodexahedron** [comment]
I just did a bunch of work on the PosTests fixture.

Coverage was increased significantly, and it revealed several bugs, most of which are noted in PosExtensions.cs or in output of tests marked Ignore (visible in test results).

The current state of that is merged to this branch.

I have not yet fixed them, as I wanted to call attention in case that raises any alarms/thoughts on your end.

**Comment by tznind** [comment]
> I just did a bunch of work on the PosTests fixture.
> 
> Coverage was increased significantly, and it revealed several bugs, most of which are noted in PosExtensions.cs or in output of tests marked Ignore (visible in test results, while allowing the fixture to "pass" for now).
> 
> The current state of that is merged to this branch.
> 
> I have not yet fixed them, as I wanted to call attention in case that raises any alarms/thoughts on your end.
> 
> Some of them are the fault of Terminal.Gui itself, but I'm pretty sure we can work around them without much work.

Thanks.  This is awesome work.  I have done a bit of manual testing and everything seems working in the editor and all the tests pass so I have merged.  Feel free to continue working on this branch or create a new one.  The diff was getting a bit unwieldy so I thought it was best to merge in.


I looked into the Pos issue and I have created a new bug in Terminal.Gui itself.  I have had to do some of this 'unpacking' of PosCombine 0 before but really it shouldn't be a thing for so simple a method:
https://github.com/gui-cs/Terminal.Gui/issues/3064

File: src/ViewFactory.cs (hunk 2 above)
Comment (dodexahedron) [reply src/ViewFactory.cs:99]: Oh yeah I forgot I was going to consolidate that. I got back to work on tests, but the good ViewFactory code is in a different branch.

I'll go back and grab that some time soon.

**Comment by dodexahedron** [comment]
Yeah, I'm cool with treating this branch as safe to merge.

I do individual work items on separate branches (one per issue, for the most part), and then merge them into this one, when done, if you ever want to look at more bite-sized chunks (or you can at least still see where they merged in, even if I've deleted the branch spec by the time you see it.

Those of course show up in the history here, basically looking like a squashed copy of all the commits on each branch into this one.

I'll keep noting in the commits and annotating in code, when I find bugs and such, unless they're simple and non-impacting, in which case I'm likely to just go ahead and take care of it in the moment, like how I went ahead and put the nullable context `?`s on those Pos Extension methods.

However, there were a couple of changes over the past few merges that are a bit bigger than nullability context:
 - KeyMap loading now uses the standard Microsoft.Extensions.Configuration API. It still uses YAMLDotNet under the hood, but now we have the ability to support any kind of configuration input supported by that, for almost no effort, in the future.
 - KeyMap and ColorSchemeBlueprint got turned into records, for a bunch of reasons, mostly centered around simpler testing and that configuration settings classes are quite well-suited for being records.
 - I have been making some internal changes mostly to private methods, largely centered around just some cleanup activities like turning big sets of if-return if-return if-return blocks into switches (preference toward expressions, for brevity, when possible) with guard clauses to cut down like 60% of the code volume for those, while also gaining an unimportant (for this kind of application) performance boost, as a side effect. Mostly, it's for readability, since many more cases can fit on the screen, making it easier for meatbags like us to deal with visually. 😅
 - Probably a few other similar things. I always make sure everything passes on existing tests before committing changes, and ensure I've got a test covering it, before merge.
 - The ViewFactory changes, which have yet to be merged, but which I want to go over with a fine-toothed comb before I do merge them, _just in case_. Shouldn't be a problem, but I'm taking a slow approach since there's no _rush_ to get there, since the current implementation does work.

I think we should do a workaround and fix for that, for a couple of reasons, regardless of if it gets fixed in Terminal.Gui. Most importantly, if it's a generic Pos, which is the only kind Terminal.Gui actually exposes publicly, any one of the IsWhateverTypeOfPos() methods will return true, if it's null. But, also, TGD has a bug in that it treats Pos types created from a literal value as absolute and zero, but then has inconsistent behavior because those methods all just return true, which breaks that assumption.

Also, I think it would be good for Terminal.Gui to actually expose the sub-classes of Pos, because they're useful to be able to identify by their type, rather than having to do what TGD already has to do, which is that string comparison of the type name. If that's "fixed" in Terminal.Gui, we wouldn't have to do that any more, which would be great. I really don't see or understand the reason for those being internal classes, over there. It creates a potential problem with inheritance, anyway, if someone were to inherit from Pos. Also, if TG exposed those types publicly, many of those extension methods would actually just plain no longer be necessary, since they're essentially a workaround for that design detail.

The only reason I didn't go ahead and do the fix to the part we have control over, on the TGD end (it would be suuuuper simple), is because that _could_ have subtle effects in the UI, if any code paths actually do reach that stuff, at runtime, and I just wanted to make you aware of it before I touched anything, in case some report were to come in related to it. I have a pretty strong feeling it does hit them, sometimes, because I've had cases when using TGD where a position mysteriously turns into another type of position, if something didn't work out quite right. Now that I've seen where that most likely comes from, I'm pretty sure fixing the null handling will, at minimum, reduce the instance of that, or at least bubble the underlying problem up to something more obvious that can be addressed, if that's not the root of it.

Alrighty. Enough rambling for now. 😅 \
I'm going to go back and make a little extra coverage pass over some fixtures I've already touched, before I move on to the next fixture, just because I'm kinda in the mood for that over the somewhat more rote conversion of existing tests. In fact, I had one more test method already in the works for PosExtensions that visual studio didn't actually save til I exited, so it didn't get included in the last commit before I merged. :|  It's got the same issue as several others in there do, around how null Pos instances get handled.

**Comment by tznind** [comment]
I think this is the PR that nuked the code coverage

image

When I run the code coverage locally I still get 75%

Maybe codeql is generating test outputs too somehow? Cant see why it is dropping to 12%

image

**Comment by tznind** [comment]
Any ideas @dodexahedron? I can't see anything that looks like it would change the CI `except .github/workflows/codeql.yml`

**Comment by tznind** [comment]
image

**Comment by tznind** [comment]
Moving this to an issue: https://github.com/gui-cs/TerminalGuiDesigner/issues/288
